kernel: activate arch interface headers

tyrantlizard77 opened the pull request:
These new headers are now pulled into the appropriate #include spaces.

Commit:
kernel: activate arch interface headers

Duplicate definitions elsewhere have been removed.

A couple functions which are defined by the arch interface
to be non-inline, but were implemented inline by native_posix
and intel64, have been moved to non-inline.

Some missing conditional compilation for z_arch_irq_offload()
has been fixed, as this is an optional feature.

Some massaging of native_posix headers to get everything
in the right scope.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

dbkinder reviewed (approved):
This documentation "harness" looks OK. I notice in the doc build artifacts (https://builds.zephyrproject.org/zephyrproject-rtos/zephyr/19807/docs/guides/porting/arch.html) that under these section headings none of the API information is showing, but as you say, renaming functions with the excluded ``z_`` prefix is coming later.

carlescufi reviewed (approved):
LGTM, thanks

ioannisg reviewed (approved):
Need to see the posix failing errors, but otherwise, legitimate.

nashif commented on doc/zephyr.doxyfile.in:
```diff
@@ -1978,6 +1978,13 @@ PREDEFINED             = "CONFIG_SYS_CLOCK_EXISTS=y" \
                          "CONFIG_BT_SMP=y" \
                          "CONFIG_USERSPACE=y" \
                          "CONFIG_BT_BREDR=y" \
+			 "CONFIG_SMP=y" \
```
fix alignment here

andrewboie replied:
I'll add this in another PR

Commit:
include: gcc.h: include header needed by posix

This header makes a call to posix_print_error_and_exit,
but pulls in no prototype.

Linker scripts use nothing in this file, move the #ifndef
_LINKER to the toplevel.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
@aescolar I had to touch some posix headers, please review.
I think there is more work to be done to fully disentangle things, but that can wait until we start work on #19666

Commit:
x86: arm: rename some functions

z_arch_ is only for those APIs in the arch interface.
Other stuff needs to be renamed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

ioannisg commented:
Can we get this merged soon? And have sufficient time until the release to test possible issues with this refactoring.

andrewboie commented:
@ioannisg waiting for @aescolar to review, I emailed him, once that's done I'll rebase for merge

aescolar commented on arch/posix/include/posix_soc_if.h:
```diff
@@ -32,7 +32,20 @@ unsigned int posix_irq_lock(void);
+static ALWAYS_INLINE void z_arch_irq_unlock(unsigned int key)
+{
+	posix_irq_unlock(key);
+}
```
How come did you move these 2 back here? (it is not a deal breaker, but I would like to understand how come you decided to move those 2 into that file. And if need be place them in a more appropriate place)

andrewboie replied:
@aescolar I moved these here because the definitions of these functions need to be where posix_irq_lock() and posix_irq_unlock() are defined.

Otherwise I have a lot of problems because of mutual dependencies. include/arch/posix/arch.h implicitly pulls in this header, but this header also implicitly pulls in include/arch/posix/arch.h due to including irq_offload.h, and since the arch interface header has prototypes to inline functions they must always be defined.

aescolar replied:
Uhm.. if you want to get this in as is, then just dismiss my review. But I would after, (myself if nobody else), untangle this.

aescolar reviewed (dismissed):
Thank you very much @andrewboie for waiting until I was back, and giving me the opportunity to review it.
I'm ok with all posix arch related changed, except for that one comment below
